File: src/cliModes.ts
// Execution engine for each CLI invocation shape (single file, scan, diff) — split out of
// cli.ts so argument parsing and mode execution are two separate, independently-readable
// concerns (the coherence detector flagged cli.ts itself for mixing both in one file).
import * as fs from 'fs';
import * as path from 'path';
import { execFileSync } from 'child_process';
const { Parser, Language } = require('web-tree-sitter');

import { analyzeSource, AnalyzeThresholds } from './core/analyze';
import { LanguageAdapter } from './core/language';
import { resolveLanguageForFile } from './languages';
import { resolveSupportedFiles } from './core/scan';
import { isIgnored, loadIgnorePatterns } from './core/esaignore';
import { FileResult, FileSummary, hasBlockingViolations, renderDiffMarkdown, renderHumanReport, renderMarkdownReport, summarize, summarizeFile, diffSummaries } from './core/report';
import { EnergyViolation, SEVERITY } from './types';

export type ReportFormat = 'json' | 'md' | 'human';

export function printUsage(): void {
    console.error('Usage: energy-state-cli <file.py|.fs|.fsx|.ts> [thresholds...]');
    console.error('       energy-state-cli <path...> [--report json|md|human] [thresholds...]              (scan a directory/subtree)');
    console.error('       energy-state-cli --base-ref <ref> [<path...>] [--report json|md] [thresholds...]  (diff PR head against a base ref)');
    console.error('Thresholds: --medium-nesting N --high-nesting N --medium-cyclomatic N --high-cyclomatic N --medium-cognitive N --high-cognitive N');
}

// decision: caches one parser instance per language adapter — scan/diff mode analyze many
// files, and re-running Parser.init()/Language.load() per file would be wasteful I/O
const parserCache = new Map<LanguageAdapter, any>();

async function loadParser(adapter: LanguageAdapter) {
    const cached = parserCache.get(adapter);
    if (cached) {
        return cached;
    }
    await Parser.init();
    const parser = new Parser();
    const grammarPath = path.join(__dirname, '..', adapter.grammarPath);
    const grammar = await Language.load(grammarPath);
    parser.setLanguage(grammar);
    parserCache.set(adapter, parser);
    return parser;
}

async function analyzeFile(filePath: string, sourceCode: string, thresholds: AnalyzeThresholds): Promise<EnergyViolation[]> {
    const adapter = resolveLanguageForFile(filePath);
    if (!adapter) {
        return [];
    }
    const parser = await loadParser(adapter);
    const tree = parser.parse(sourceCode);
    return analyzeSource(sourceCode, tree, adapter, filePath, thresholds);
}

function exitForCounts(counts: { low: number; medium: number; high: number }): never {
    process.exit(hasBlockingViolations(counts) ? 1 : 0);
}

export async function runLegacySingleFile(filePath: string, thresholds: AnalyzeThresholds): Promise<void> {
    const adapter = resolveLanguageForFile(filePath);
    if (!adapter) {
        console.error(`Unsupported file type: ${filePath}`);
        printUsage();
        process.exit(2);
    }

    const sourceCode = fs.readFileSync(filePath, 'utf8');
    const violations = await analyzeFile(filePath, sourceCode, thresholds);

    console.log(JSON.stringify(violations, null, 2));

    const hasBlockingViolation = violations.some(v => v.severity === SEVERITY.HIGH || v.severity === SEVERITY.MEDIUM);
    process.exit(hasBlockingViolation ? 1 : 0);
}

export async function runScan(paths: string[], thresholds: AnalyzeThresholds, reportFormat: ReportFormat): Promise<void> {
    const files = resolveSupportedFiles(paths);
    const results: FileResult[] = [];

    for (const filePath of files) {
        const sourceCode = fs.readFileSync(filePath, 'utf8');
        const violations = await analyzeFile(filePath, sourceCode, thresholds);
        // decision: reports the path relative to cwd — resolveSupportedFiles resolves
        // absolute paths internally for reliable dedup, but a repo-wide report reads far
        // better as "src/foo.ts" than a full filesystem path repeated on every row
        results.push({ filePath: path.relative(process.cwd(), filePath), violations });
    }

    const summary = summarize(results);
    if (reportFormat === 'human') {
        console.log(renderHumanReport(results));
    } else if (reportFormat === 'md') {
        console.log(renderMarkdownReport(summary));
    } else {
        console.log(JSON.stringify(summary, null, 2));
    }
    exitForCounts(summary.totalCounts);
}

function changedFilesFromGit(baseRef: string): string[] {
    const output = execFileSync('git', ['diff', '--name-only', '--diff-filter=d', `${baseRef}...HEAD`], { encoding: 'utf8' });
    return output.split('\n').map(line => line.trim()).filter(Boolean);
}

// decision: returns null (not throwing) when the file doesn't exist at baseRef — a newly
// added file has no base content, which is a normal "new file" case, not an error
function readAtRef(ref: string, filePath: string): string | null {
    try {
        // decision: pipes stderr instead of inheriting it — git's own "fatal: path ... exists
        // on disk, but not in <ref>" would otherwise print to the CLI's stderr on every new
        // file, duplicating (and outnumbering) our own, already-informative message below
        return execFileSync('git', ['show', `${ref}:${filePath}`], { encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'] });
    } catch (error) {
        console.error(`energy-state-cli: could not read ${filePath} at ${ref} (new file or rename) — treating as new`);
        return null;
    }
}

export async function runDiff(baseRef: string, explicitPaths: string[], thresholds: AnalyzeThresholds, reportFormat: ReportFormat): Promise<void> {
    const rootDir = process.cwd();
    const ignorePatterns = loadIgnorePatterns(rootDir);
    const changedFiles = (explicitPaths.length > 0 ? explicitPaths : changedFilesFromGit(baseRef))
        .filter(filePath => resolveLanguageForFile(filePath) && fs.existsSync(filePath))
        .filter(filePath => !isIgnored(path.resolve(filePath), rootDir, ignorePatterns));

    const headSummaries: FileSummary[] = [];
    const baseSummaries: FileSummary[] = [];

    for (const filePath of changedFiles) {
        const headSource = fs.readFileSync(filePath, 'utf8');
        const headViolations = await analyzeFile(filePath, headSource, thresholds);
        headSummaries.push(summarizeFile({ filePath, violations: headViolations }));

        const baseSource = readAtRef(baseRef, filePath);
        if (baseSource !== null) {
            const baseViolations = await analyzeFile(filePath, baseSource, thresholds);
            baseSummaries.push(summarizeFile({ filePath, violations: baseViolations }));
        }
    }

    const entries = diffSummaries(baseSummaries, headSummaries);
    console.log(reportFormat === 'md' ? renderDiffMarkdown(entries, baseRef) : JSON.stringify(entries, null, 2));

    // decision: diff-mode's exit code reflects whether anything in the diff *worsened*, not
    // "does head have any medium/high violation" (scan/legacy mode's rule) — a PR touching a
    // file that already carried debt, or a genuinely new file, should not fail this check on
    // pre-existing severity alone; only a file this PR made worse should block it
    process.exit(entries.some(entry => entry.status === 'worsened') ? 1 : 0);
}

File: src/CliModes.fs
module Energy.CliModes

open System.Threading.Tasks

open Fable.Core.JsInterop

open Energy.CliNode
open Energy.CliRuntime
open Energy.Core.Analyze
open Energy.Core.Esaignore
open Energy.Core.Paths
open Energy.Core.Report
open Energy.Core.ReportDiff
open Energy.Core.ReportHuman
open Energy.Core.Scan
open Energy.Core.Violation
open Energy.Languages.Registry

// decision: the report format is a closed set of CLI choices, so it is a DU rather than a free
// string — this is what keeps the render dispatch below from being a stringly-typed match (the
// primitive-obsession detector's stringly-control-flow check) and makes the selection exhaustive at
// compile time. The `--report` flag still takes a string; parseReportFormat maps it onto the DU.
type ReportFormat =
    | Human
    | Markdown
    | Sarif
    | Json

let parseReportFormat (text: string) : ReportFormat =
    match text with
    | "human" -> Human
    | "md"
    | "markdown" -> Markdown
    | "sarif" -> Sarif
    | _ -> Json

let printUsage () =
    error "Usage: energy-state-cli <file.py|.fs|.fsx|.ts> [thresholds...]"

    error
        "       energy-state-cli <path...> [--report json|md|human|sarif] [thresholds...]        (scan a directory/subtree)"

    error
        "       energy-state-cli --base-ref <ref> [<path...>] [--report json|md] [thresholds...]  (diff PR head against a base ref)"

    error
        "Thresholds: --medium-nesting N --high-nesting N --medium-cyclomatic N --high-cyclomatic N --medium-cognitive N --high-cognitive N --medium-parameter-count N --high-parameter-count N"

    error "Flags: --include-test-files (also flag magic numbers and magic strings in test files)"

let private violationJson violation =
    let hotspots =
        violation.Hotspots
        |> List.map (fun hotspot -> createObj [ "line" ==> hotspot.Line; "weight" ==> hotspot.Weight ])
        |> List.toArray

    createObj
        [ "line" ==> violation.Line
          "column" ==> violation.Column
          "type" ==> violationTypeName violation.Type
          "severity" ==> severityName violation.Severity
          "message" ==> violation.Message
          "hotspots" ==> hotspots ]

// decision: composes JSON report rows from FileResult rather than FileSummary so agent consumers
// receive the same location and remediation message as VS Code without changing summary consumers.
let summaryJson results =
    let summary = summarize results

    let files =
        results
        |> List.map (fun result ->
            let file = summarizeFile result

            createObj
                [ "filePath" ==> file.FilePath
                  "score" ==> file.Score
                  "counts"
                  ==> createObj
                          [ "low" ==> file.Counts.Low
                            "medium" ==> file.Counts.Medium
                            "high" ==> file.Counts.High ]
                  "byType"
                  ==> (file.ByType
                       |> Map.toList
                       |> List.map (fun (key, value) -> key ==> value)
                       |> createObj)
                  "violations" ==> (result.Violations |> List.map violationJson |> List.toArray) ])
        |> List.toArray

    createObj
        [ "files" ==> files
          "totalScore" ==> summary.TotalScore
          "totalCounts"
          ==> createObj
                  [ "low" ==> summary.TotalCounts.Low
                    "medium" ==> summary.TotalCounts.Medium
                    "high" ==> summary.TotalCounts.High ] ]

let private diffJson entries =
    entries
    |> List.map (fun entry ->
        let status =
            match entry.Status with
            | New -> "new"
            | Improved -> "improved"
            | Worsened -> "worsened"
            | Unchanged -> "unchanged"

        createObj
            [ "filePath" ==> entry.FilePath
              "baseScore" ==> entry.BaseScore
              "headScore" ==> entry.HeadScore
              "delta" ==> entry.Delta
              "status" ==> status ])
    |> List.toArray
    |> box

let runScan (paths: string list) (thresholds: AnalyzeThresholds) (reportFormat: ReportFormat) : Task<unit> =
    task {
        let! analysis = analyzeFiles (resolveSupportedFiles paths (cwd ())) thresholds

        match analysis with
        | Error analysisError ->
            error ("energy-state-cli failed: " + analysisErrorMessage analysisError)
            exit 1
        | Ok results ->
            let summary = summarize results

            output (
                match reportFormat with
                | Human -> renderHumanReport results
                | Markdown -> renderMarkdownReport summary
                | Sarif -> stringify (Energy.Core.ReportSarif.renderSarif results)
                | Json -> stringify (summaryJson results)
            )

            exit (if hasBlockingViolations summary.TotalCounts then 1 else 0)
    }

let private changedFilesFromGit baseRef =
    execFileSync
        "git"
        [| "diff"; "--name-only"; "--diff-filter=d"; baseRef + "...HEAD" |]
        (createObj [ "encoding" ==> "utf8" ])
    |> fun result ->
        result.Split('\n')
        |> Array.map _.Trim()
        |> Array.filter ((<>) "")
        |> Array.toList

// decision: a missing base version is a normal newly-added/renamed file, not a failed analysis;
// git's own stderr is intentionally suppressed so one concise explanatory line is emitted.
let private readAtRef reference filePath =
    try
        Some(
            execFileSync
                "git"
                [| "show"; reference + ":" + filePath |]
                (createObj [ "encoding" ==> "utf8"; "stdio" ==> [| "ignore"; "pipe"; "ignore" |] ])
        )
    with _ ->
        error (
            "energy-state-cli: could not read "
            + filePath
            + " at "
            + reference
            + " (new file or rename) — treating as new"
        )

        None

let runDiff
    (baseRef: string)
    (explicitPaths: string list)
    (thresholds: AnalyzeThresholds)
    (reportFormat: ReportFormat)
    : Task<unit> =
    task {
        let rootDir = cwd ()
        let patterns = loadIgnorePatterns rootDir

        let changed =
            (if explicitPaths.IsEmpty then
                 changedFilesFromGit baseRef
             else
                 explicitPaths)
            |> List.filter (fun filePath ->
                resolveLanguageForFile filePath |> Option.isSome && existsSync (Path filePath))
            |> List.filter (fun filePath -> not (isIgnored (resolvePath (Path filePath)) (Path rootDir) patterns))

        let rec analyzeChanged files bases heads =
            task {
                match files with
                | [] -> return Ok(List.rev bases, List.rev heads)
                | filePath :: remaining ->
                    let! headAnalysis = analyzePath (Path filePath) thresholds

                    match headAnalysis with
                    | Error analysisError -> return Error analysisError
                    | Ok headViolations ->
                        let head =
                            summarizeFile
                                { FilePath = filePath
                                  Violations = headViolations }

                        match readAtRef baseRef filePath with
                        | None -> return! analyzeChanged remaining bases (head :: heads)
                        | Some baseSource ->
                            let! baseAnalysis = analyzeFile (Path filePath) baseSource thresholds

                            match baseAnalysis with
                            | Error analysisError -> return Error analysisError
                            | Ok baseViolations ->
                                let baseSummary =
                                    summarizeFile
                                        { FilePath = filePath
                                          Violations = baseViolations }

                                return! analyzeChanged remaining (baseSummary :: bases) (head :: heads)
            }

        let! analysis = analyzeChanged changed [] []

        match analysis with
        | Error analysisError ->
            error ("energy-state-cli failed: " + analysisErrorMessage analysisError)
            exit 1
        | Ok(bases, heads) ->
            let entries = diffSummaries bases heads

            output (
                if reportFormat = Markdown then
                    renderDiffMarkdown entries baseRef
                else
                    stringify (diffJson entries)
            )

            // invariant: diff mode blocks only regressions; existing debt and newly added files are
            // reported but do not fail a PR until their score worsens relative to the base revision.
            exit (
                if entries |> List.exists (fun entry -> entry.Status = Worsened) then
                    1
                else
                    0
            )
    }
